pallets/creditcoin/src/types/collect_coins.rs

```diff
@@ -0,0 +1,120 @@
+use super::*;
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct CollectedCoins<Hash, Balance> {
+	pub to: AddressId<Hash>,
+	pub amount: Balance,
+	pub tx_id: ExternalTxId,
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct UnverifiedCollectedCoins {
+	pub to: ExternalAddress,
+	pub tx_id: ExternalTxId,
+	pub contract: GCreContract,
+}
+
+impl UnverifiedCollectedCoins {
+	pub fn into_output<T>(self, amount: T::Balance) -> CollectedCoins<T::Hash, T::Balance>
+	where
+		T: Config,
+	{
+		let Self { to, tx_id, contract: GCreContract { chain, .. } } = self;
+		let to = crate::AddressId::new::<T>(&chain, to.as_slice());
+		CollectedCoins { amount, to, tx_id }
+	}
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct CollectedCoinsId<Hash>(Hash);
+
+impl<H> CollectedCoinsId<H> {
+	fn inner_hash<Hasher>(blockchain: &Blockchain, blockchain_tx_id: &[u8]) -> H
+	where
+		Hasher: Hash<Output = H>,
+	{
+		let key = concatenate!(&*blockchain.as_bytes(), blockchain_tx_id);
+		<Hasher as Hash>::hash(&key)
+	}
+
+	pub fn new<C: SystemConfig>(
+		contract_chain: &Blockchain,
+		blockchain_tx_id: &[u8],
+	) -> CollectedCoinsId<H>
+	where
+		<C as SystemConfig>::Hashing: Hash<Output = H>,
+	{
+		let hash = Self::inner_hash::<C::Hashing>(contract_chain, blockchain_tx_id);
+		CollectedCoinsId(hash)
+	}
+}
+
+impl<H> From<H> for CollectedCoinsId<H> {
```

Comment:
Same here, can we put a `#[cfg(test)]` on this?

Reply:
Same arguments, `CollectedCoinsId::from` is needed in a few places here to go from TaskId (i.e. config::Hash) -> to CollectedCoinsId. Everything is encapsulated in the TaskV2 implementation so it should be impossible to misuse From<T::Hash> if you implement the interface correctly which is pretty straight forward imo. Do not From<Hash> -> ItemId<Hash> if you didn't generate the hash from <'pending task'  as TaskV2>::to_id(..) and you are good, just don't mix Hashes which don't shouldn't happen if you respect the interfaces.

```rust
impl<T: Config> TaskV2<T> for UnverifiedCollectedCoins
where
	UnverifiedCollectedCoins: OffchainVerification<T>,
	<UnverifiedCollectedCoins as OffchainVerification<T>>::Output: Into<T::Balance>,
{
	type Call = crate::pallet::Call<T>;
	type EvaluationError = VerificationFailureCause;
	type SchedulerError = SchedulerError;
	fn to_id(&self) -> T::Hash {
		CollectedCoinsId::inner_hash::<T::Hashing>(&self.contract.chain, self.tx_id.as_slice())
	}

	fn persistence_call(
		&self,
		deadline: T::BlockNumber,
		id: &T::Hash,
	) -> Result<Self::Call, TaskError<Self::EvaluationError, Self::SchedulerError>> {
		use crate::ocw::OffchainError::*;
		match self.verify() {
			Ok(amount) => {
				let coins = self.clone().into_output::<T>(amount.into());
				let id = CollectedCoinsId::from(*id);
				Ok(Self::Call::persist_task_output {
					deadline,
					task_output: TaskOutput::from((id, coins)),
				})
			},
			Err(InvalidTask(cause)) if cause.is_fatal() => {
				log::warn!("Failed to verify pending task {:?} : {:?}", self, cause);
				let id = CollectedCoinsId::from(*id);
				Ok(Self::Call::fail_task { deadline, task_id: id.into(), cause })
			},
			Err(InvalidTask(e)) => Err(TaskError::Evaluation(e)),
			Err(NoRpcUrl(e)) => Err(TaskError::Scheduler(e.into())),
			Err(RpcError(e)) => Err(TaskError::Scheduler(e.into())),
			Err(IncorrectChainId) => Err(TaskError::Scheduler(SchedulerError::IncorrectChainId)),
		}
	}

	fn is_persisted(id: &T::Hash) -> bool {
		let id = CollectedCoinsId::from(*id);
		crate::pallet::CollectedCoins::<T>::contains_key(&id)
	}
}
```